File: server/strategies/affine.py
from .cipher_interface import CipherInterface

class AffineCipher(CipherInterface):
    def mod_inverse(self, a, m):
        for x in range(1, m):
            if (a * x) % m == 1: return x
        return None

    def decrypt(self, text: str, key: str) -> str:
        try:
            # Anahtar formatı: "5,8" (a=5, b=8)
            parts = key.split(',')
            a, b = int(parts[0]), int(parts[1])
            
            a_inv = self.mod_inverse(a, 26)
            if a_inv is None: return "Hata: 'a' sayısı 26 ile aralarında asal olmalı."

            result = ""
            for char in text:
                if char.isalpha():
                    y = ord(char.upper()) - 65
                    # D(x) = a^-1 * (y - b) % 26
                    decrypted_char = chr(((a_inv * (y - b)) % 26) + 65)
                    result += decrypted_char if char.isupper() else decrypted_char.lower()
                else:
                    result += char
            return result
        except:
            return "Hata: Anahtar formatı 'a,b' olmalı (Örn: 5,8)"

File: lib/ciphers/affine.dart
import 'cipher_base.dart';

class AffineCipher implements CipherBase {
  @override
  String get name => "Affine (Doğrusal)";
  @override
  String get id => "affine";

  int gcd(int a, int b) => b == 0 ? a : gcd(b, a % b);

  @override
  String encrypt(String text, String key) {
    // Key format: "5,8"
    var parts = key.split(',');
    if (parts.length != 2) return "Anahtar 'a,b' formatında olmalı (Örn: 5,8)";

    int a = int.tryParse(parts[0]) ?? 0;
    int b = int.tryParse(parts[1]) ?? 0;

    if (gcd(a, 26) != 1) return "Hata: 'a' sayısı 26 ile aralarında asal olmalı.";

    StringBuffer result = StringBuffer();
    for (int i = 0; i < text.length; i++) {
      int charCode = text.codeUnitAt(i);
      if (charCode >= 65 && charCode <= 90) {
        result.write(String.fromCharCode(((a * (charCode - 65) + b) % 26) + 65));
      } else if (charCode >= 97 && charCode <= 122) {
        result.write(String.fromCharCode(((a * (charCode - 97) + b) % 26) + 97));
      } else {
        result.write(text[i]);
      }
    }
    return result.toString();
  }
}